### PR TITLE
Allow flooding of farming crops

### PR DESCRIPTION
**Pull request by rubenwardy**
This allows harvesting crops using water. It's not realistic, but is very useful

### CODE REVIEW

**Comment by Fixer-007** [comment]
In game physics opens many possibilities, this change is one of them. Minetest is very static and has almost zero interactivity apart from fire/water. This change is the first move to expand it.

**Comment by paramat** [comment]
As discussed before in issues, i'm opposed, but it looks like i'm outvoted :-1: 

If MC didn't have this feature, it wouldn't be requested or contributed, as it doesn't make much sense and is very unintuitive.
I believe that most of the reason some people think this is a good idea is because MC has made it seem normal (not suggesting this is the case for the author).
This will create an opportunity for very easy griefing of crops, which was one of the concerns discussed before when floodable plants was discussed. Also the creation of much more mess when accidental spills happen. Currently, luckily, liquids don't destroy much.
Harvesting by flooding is indeed not realistic at all, as admitted by the author, so forgetting about MC, it seems bizarre. MTG farming is loosely based on realism in all other ways.

I'm not sure servers will be happy about this, please can we get some more server admin input? @Ezhh @VanessaE @shivajiva @tenplus1

**Comment by paramat** [comment]
If merged, please could this be optional and switched by a setting? It may need to be default off to avoid problems.

**Comment by LoneWolfHT** [comment]
> I believe that most of the reason some people think this is a good idea is because MC has made it seem normal (not suggesting this is the case for the author).

If MC came up with it I'm pretty sure we could have too

> This will create an opportunity for very easy griefing of crops, which was one of the concerns discussed before when floodable plants was discussed.

Most sensible server owners prevent placement of liquids above y 0 without a special privilege

> Also the creation of much more mess when accidental spills happen.

If you mean when accidentally spilling on a crop: I don't get your point

> If merged, please could this be optional and switched by a setting? It may need to be default off to avoid problems.

A setting sounds good. I'm neutral on the default

**Comment by rubenwardy** [comment]
I had no idea that MC had this feature

**Comment by paramat** [comment]
rubenwardy, LOL ok =)

My point is, currently, an accidental spill of liquid on a crop does not drop the crops.
Anyway, if optional and default disabled i'll be neutral, some will like it.

**Comment by benrob0329** [comment]
> So it seems most players won't be able to use this feature?

Just place the farm a block lower

**Comment by LoneWolfHT** [comment]
> LoneWolfHT, my point is, currently, an accidental spill of liquid on a crop does not drop the crops.

That sounds fine to me. It's kind of like swords, or even pickaxes. You can kill but you can also be killed. You can mine underground but you can also accidentally mine into lava, and many other things...

> So it seems most players won't be able to use this feature?

On the server I played on trusted players & staff members would get the priv. Then the players that didn't have the priv would just ask the players/staff with the priv to place their liquids for them

**Comment by jastevenson303** [comment]
I've beeñ doing this on lavaland and it's pretty fun, adds a lot.

I also did not know Minecraft did this, paramat.  Is it important?

File: mods/farming/api.lua [hunk 1]
```diff
@@ -370,6 +370,14 @@ farming.register_plant = function(name, def)
 			on_timer = farming.grow_plant,
 			minlight = def.minlight,
 			maxlight = def.maxlight,
+			floodable = true,
+			on_flood = function(pos, oldnode, newnode)
+				local items = minetest.get_node_drops(oldnode.name, nil)
+				for j=1, #items do
```
Comment (DS-Minetest) [inline mods/farming/api.lua]:
```suggestion
				for j = 1, #items do
```

**Comment by DS-Minetest** [comment]
- I think, the majority of servers allow placing water (and river water) above y=0.
- Maybe the items shouldn't be dropped if the flooding liquid (newnode) is lava.
- It is possible that this change isn't very realistic, but it feels far more realistic than shrubbery (and similar, all that stuff that contains air and isn't too solid) blocking water from flowing.

**Comment by paramat** [comment]
Instead, how about making the grown crops 'dig immediate' instead of 'snappy', then harvesting would be as quick as walking non-stop along rows, and would not be much more time and effort than picking up many dropped objects.
'Smappy' does seem quite a lot of effort for harvesting, 'dig immediate' seems suitable for crops.

The farming mod is loosely based on reality, but in reality floods destroy crops, so adding something completely opposed to reality seems irrational. It makes as much sense as harvesting using TNT or fire. It would at least make some sense if flooding destroyed the nodes, with no drops (but to be clear i am not suggesting this).

'Not realistic but useful' could justify anything, might as well have the crop pop up immediately after planting the seed =) The point of farming is the loose realism and the effort needed, which improves the game.

**Comment by Fixer-007** [comment]
The reason MC has it not just simply because it has it, but because it adds to gameplay, that is why the game is so popular, it has a lot of good mechanics/interactivity that allows you to do a lot of stuff, like farming those crops by water streams that carry them down the hole to a chest, that is like year 2011 technology (8 years ago, mind you). Stuff like entity bouncing from each other, stream dynamics, and other stuff opens huge possibilities that are still impossible in minetest game.

**Comment by rubenwardy** [comment]
My usecase it to make semi-automated farms by allowing the field to be harvested:

unknown1

Farming is currently very work intensive. Below the fields I have pipeworks vacuum tubes which pipe up dropped items, but it items followed flowing water then that would work in vanilla MTG

**Comment by rubenwardy** [comment]
I don't like limiting the gameplay of singleplayer because of borderline griefing that may occur on multiplayer. Crops don't tend to be that valuable, and water can already do a lot of damage on servers that have drowning

This could be extended to not work in protected areas

**Comment by mbartlett21** [comment]
> The farming mod is loosely based on reality, but in reality **floods destroy crops**, so adding something completely opposed to reality seems irrational. [...] It would at least make some sense if **flooding destroyed the nodes, with no drops** (but to be clear i am not suggesting this).

Maybe make a setting that every flooded crop has a chance of 1 in N to drop itself?

**Comment by DS-Minetest** [comment]
>This could be extended to not work in protected areas

That would need different nodes with `floodable = false`, eh.

**Comment by rubenwardy** [comment]
> That would need different nodes with floodable = false, eh.

If you return true in on_flood then the flood is cancelled. This results in repeated calls to on_flood, however

**Comment by paramat** [comment]
MC doing this not being important is actually my point =) i am concerned people will be more likely to think this is rational and acceptable because MC has made it seem normal. MC does some silly things.

Fixer-007, the way i see it is: Interactivity can be added in other ways. A bad idea that adds interactivity is not justified just by adding interactivity, which is what your posts seem to imply.
MT can be popular by adding interactivity but without copying MC features, i'd prefer it adds features that are loosely based on reality as farming is.

Maybe we can come up with a better automated farming idea? Flooding crops to drop them as objects is sloppy and very unrealistic.

It seems to be a feature primarily justified by allowing automation, this is probably why MC added the feature.
Please can we have a setting for this? I can understand why the default will probably be preferred as 'enabled' so i won't insist otherwise.

**Comment by paramat** [comment]
Hmm ... i'm guessing that this happened in MC because first they had items moving in flowing liquid, and they then realised if liquid dropped crops it would provide a simple way to harvest and transport them. I guess they decided this justified the anti-realism.
But MT doesn't have items moving in flowing liquid (yet), so it's not as obvious a thing to do.

If the justification is automation of farming (which is good to have i agree), i think it is the responsibility of the automation mod to provide a feature that drops the crops, maybe a machine, something better than using flooding to do so. Most farming is not automated so i would prefer the 'dropping' feature isn't a part of the farming mod.

**Comment by jastevenson303** [comment]
Not everyone has the same expectations of what a video game should be,
paramat.  Personally I think your "refined taste" could use some seasoning.

It's a video /game/.  It should be /fun/.

**Comment by DS-Minetest** [comment]
Imo this feature shouldn't be added because of automatisation, but because it feels more realistic. Crops blocking water is so weird.
And could we please stop writing about that other game and instead discuss about this feature as it will be in minetest?

**Comment by TumeniNodes** [comment]
Flooding crops should kill off 1/2 the crops, thus making farming harder... not making farming easier.
 In my mind _this_ would make game-play more interesting, because it would cause a need for the player to find a solution. But in real life, this is not the way people think today.

 Those of us who think more "reality based mechanics", need to realize that end users are more pleased when things are made easier for them to achieve goals in both gaming life and real life. No matter how unrealistic.
 It seems to me, that people are more satisfied with games... which do not challenge them in any way, but rather make everything easier to win/succeed. : /

**Comment by LoneWolfHT** [comment]
> It seems to me, that people are more satisfied with games... which do not challenge them in any way, but rather make everything easier to win/succeed. : /

All the servers would be creative servers if this were completely true.
Farming crops is more monotonous than a challenge. It's just dig, plant, wait, repeat.

**Comment by rubenwardy** [comment]
Farming became popular because it's more efficient, and frees up time to do other things. Farming in vanilla MTG is time intensive

**Comment by Df458** [comment]
> It seems to me, that people are more satisfied with games... which do not challenge them in any way, but rather make everything easier to win/succeed. : /

I think that declaration misses the point of the feature. I'm not about to argue for or against the PR itself, but given a choice I believe that water harvesting crops is consistently _more interesting_ than water killing them. I think it's pretty obvious if you step back and think about the context and results.

Under normal circumstances, water is unlikely to be above or at the same level as crops. The only time someone would deliberately set things up this way is if their farm was underground (or otherwise below sea level, like an underwater base) or set up vertically. If water touching plants has no benefit, it will almost never happen **except by accident**. In other words, it's not a challenge for a normal player, just a setback that happens completely by accident. This is exactly the kind of thing that frustrates people, because it comes out of nowhere.

Now, let's consider the repercussions of harvesting with water. This definitely affects game balance, since it makes harvesting faster and easier. _However_, it also introduces a fun kind of challenge: Designing a system that, given the current fluid mechanics, can harvest plants efficiently without destroying anything else (such as torches, a pretty common light source). It's not particularly difficult, but it makes the player think about how to use the mechanic instead of punishing them for... what? Accidentally right-clicking? Breaking the wrong node?

Looking through this sort of lens, one side is an unlikely punishment and the other is a mechanic that players will choose to engage with (because there's something in it for them). If water killed plants but also had **a good reason to be above them in the first place**, then I think these options would be evenly balanced. But as it is, I don't see a good reason to implement the mechanic this way.

> Flooding crops should kill off 1/2 the crops

Frankly, this is even worse than the above. If a new player (trying this because they've seen or played another similar game before) gets lucky, they may be led to believe that the method works. Then they'll get punished for it in the future, and it will be entirely the RNG's fault.

With all that said, I also think cherry-picking individual mechanics like this has its own problems. I think one of the reasons "the other game" is so much more engaging is that it's designed more in terms of large interlocking systems, rather than separate interactions. The former lends itself better to emergence, and with it creative problem-solving and a stronger feeling of immersion.

**Comment by TumeniNodes** [comment]
>Those of us who think more "reality based mechanics", need to realize...

I made that statement, meaning those of us who have a problem with such a mechanic due to lack of reality, should consider that it would be preferred by end users.
I did not make my comment as being against this.
My attempt to point this out was ` "flooded" ` out by my side-rant.

Anything which will attract, and keep players engaged is +1 with me.

**Pull request closed without merging**